### DISCUSSION TITLE
Is the `enhance` phase essential? Is there documentation on what it does?

### DISCUSSION THREAD

**Discussion by markstos** [valhalla/valhalla#3744] [Q&A] [answered]
I'm experiencing a hang during the `enhance` phase, the same as described here:

https://github.com/valhalla/valhalla/issues/3499

In that case, the issue was blamed on bad OSM data and closed, so it could recur. I'm running into building using the Austria and Switzerland extracts together.

One solution I'm considering is to just skip the "enhance" phase, but I'm having trouble finding documentation on what it does and if it's essential.  I searched the documentation but didn't find any clear result:

https://valhalla.readthedocs.io/en/latest/search.html?q=enhance

**_Is the `enhance` phase essential? Where I find documentation on what it does?_**

**Comment by ImreSamu** [comment]
> 'm running into building using the Austria and Switzerland extracts together.

imho: it is better to cut the area from a larger region ( from https://download.geofabrik.de/europe/dach.html )
* so no data integrity problem;
      * no missing ways / roads.
      * no different way versions.
* and no missing areas (like [Liechtenstein](https://download.geofabrik.de/europe/liechtenstein.html) 

For Creating geographic extracts: 
   * see:  https://osmcode.org/osmium-tool/manual.html#creating-geographic-extracts

IMHO: this is a more straightforward and fail-safe method.

**Reply by markstos** [reply]
@ImreSamu It turns out I need Germany too, so this is a great suggestion! I'll give it a try.

**Reply by markstos** [reply]
If I use two extracts for two adjacent regions instead of one extract for both, does that mean I could end up with routing problems between the two regions because some details along the border are corrupted?

**Reply by ImreSamu** [reply]
>I could end up with routing problems between the two regions because some details along the border are corrupted?

not always, but it can happen.

This can be a problem especially if you want to treat extracts on different time bases.
So never use the "austria-latest.osm.pbf" + "switzerland-latest.osm.pbf" together ..
* it is better to use timestamped version:  "austria-220911.osm.pbf"  + "switzerland-220911.osm.pbf"

And you have to check the borderline ...  
* any overlap ? ( There is always some overlap ) 
   *  So the osm data should be the same version.
* missing part?  ( yes..  in this case : Lichenstein ) 

![image](https://user-images.githubusercontent.com/2217320/189773475-7ff35606-626c-41e7-9463-a76188f3097d.png)

Any you have to check the Strategies used to create your input extract ( "simple"?, "complete_ways"? or "smart"? )  
* https://github.com/osmcode/osmium-tool/blob/master/man/osmium-extract.md#strategies

In Europe it may even be important to leave a +300km buffer above the border, as borders are rarely straight, and sometimes the quickest way to reach a settlement is via another country ( and in the European Union the borders are open )
 
 For example Geneve (Swiss)  -> Lugarno(Swiss ) .. ( via France .. Italy  )
![image](https://user-images.githubusercontent.com/2217320/189775021-4c79bb8c-4b4e-40f4-ab03-b12400737928.png)

**Reply by markstos** [reply]
Fascinating and informative @ImreSamu! Köszönöm! 

I think if I need more than the Austrian/Switzerland/Germany extracts for this project I'll just try the whole Europe extract and see how many resources that consumes. 😂

**Comment by nilsnolde** [comment, marked as answer]
We currently have a planet build with OSM data updated yesterday. Should be in `enhance` stage later today, I'll update if it finished for us.

The enhance stage is technically optional, you could just stop graph building after the initial tiles are built. BUT, that'll leave with a fairly broken graph for how valhalla traverses it. I can't tell you all the details the `enhance` stage does, but valhalla will not be fully functional with all request parameters (e.g. avoid country borders). Also you won't be able to build hierarchies (correct me if I'm wrong..), meaning you won't really be able to process routes longer than a few dozen kms without severe performance penalties.

**Reply by markstos** [reply]
Thanks! When I tried building with the DACH extract, that passed the `enhance` stage without a problem.

**Comment by kevinkreiser** [comment]
Enhance is definitely essential. The graph will not be properly connected without it nor will many essential properties be set. To avoid issues with data you can simply download the planet and use osmium to cut out a polygonal region that includes AT CH and LI as a step before starting your tile build
